ComponentKit/Core/CKBuildComponent.mm

```diff
@@ -28,6 +66,13 @@ CKBuildComponentResult CKBuildComponent(CKComponentScopeRoot *previousRoot,
   CKThreadLocalComponentScope threadScope(previousRoot, stateUpdates);
   // Order of operations matters, so first store into locals and then return a struct.
   CKComponent *const component = componentFactory();
+  // Build the component tree from the render function.
+  CKBuildComponentTree(component,
+                       threadScope.newScopeRoot.rootNode,
+                       previousRoot.rootNode,
+                       threadScope.newScopeRoot,
+                       stateUpdates);
```

Comment:
I see — it seems that you only call `CKBuildComponentTree` exactly once, immediately after render, before anyone else has a reference to the given component.

One of the principles I think it's worth striving for: data structures should be completely immutable after their creation. I am not sure if that will be possible in this particular case, but I think it would be useful to try.

Reply:
I agree.